File: FTP Client for WinForms using Ultimate FTP library/[VB]-FTP Client for WinForms using Ultimate FTP library/VB/Ftp/FtpClientPlugin.vb
Imports System.Security.Cryptography.X509Certificates
Imports System.Text
Imports ClientSample.Ftp.Security
Imports ComponentPro
Imports ComponentPro.IO
Imports ComponentPro.Net
Imports ComponentPro.Security.Certificates

Namespace ClientSample.Ftp
	Friend Class FtpClientPlugin
		Implements IClientPlugin
		Private _controller As ClientController
		Private _client As ComponentPro.Net.Ftp
		Private _view As IClientView
		Private _loginSettings As SettingInfoBase

		Public Sub New(ByVal view As IClientView, ByVal clientController As ClientController, ByVal loginSettings As SettingInfoBase)
			_view = view
			_controller = clientController
			_loginSettings = loginSettings
		End Sub

		Public Function Create() As IRemoteFileSystem Implements IClientPlugin.Create
			Dim client As New ComponentPro.Net.Ftp()

#If Not Framework4_5 Then
			AddHandler client.SetMultipleFilesPermissionsCompleted, AddressOf client_SetMultipleFilesPermissionsCompleted
			AddHandler client.UploadUniqueFileCompleted, AddressOf client_UploadUniqueFileCompleted
			AddHandler client.CertificateRequired, AddressOf client_CertificateRequired
			AddHandler client.CertificateReceived, AddressOf client_CertificateReceived
#End If

			_client = client

			Return _client
		End Function

		''' <summary>
		''' Handles the client's SetMultipleFilesPermissionsCompleted event.
		''' </summary>
		''' <param name="sender">The Ftp object.</param>
		''' <param name="e">The event arguments.</param>
		Private Sub client_SetMultipleFilesPermissionsCompleted(ByVal sender As Object, ByVal e As ExtendedAsyncCompletedEventArgs(Of FileSystemTransferStatistics))
			If e.Error IsNot Nothing Then
				If Not _controller.HandleException(e.Error) Then
					Return
				End If
			End If

			_controller.RefreshRemoteList()
			_controller.EnableProgress(False)
		End Sub

		Public Function GetPermissionsString(ByVal f As AbstractFileInfo) As String Implements IClientPlugin.GetPermissionsString
			If CType(f, FtpFileInfo).System = FtpFileOs.Unix Then
				Return ToPermissions(CType(f, FtpFileInfo).Permissions)
			Else
				Return "-"
			End If
		End Function

		Private Sub IClientPlugin_AuthenticatePost() Implements IClientPlugin.AuthenticatePost
			If _loginSettings.Get(Of Boolean)(FtpLoginInfo.ClearCommandChannel) AndAlso _loginSettings.Get(Of FtpSecurityMode)(FtpLoginInfo.SecurityMode) <> FtpSecurityMode.None Then
				CType(_client, ComponentPro.Net.Ftp).ClearCommandChannel()
			End If
		End Sub

		Public Function CanBeReconnected(ByVal exc As Exception) As Boolean Implements IClientPlugin.CanBeReconnected
			Dim fe As FtpException = TryCast(exc, FtpException)
			Dim se As System.Net.Sockets.SocketException

			If fe IsNot Nothing Then
				If fe.Status = FtpExceptionStatus.ConnectionClosed Then
					Return True
				End If
			Else
				se = TryCast(exc, System.Net.Sockets.SocketException)
				If (se IsNot Nothing AndAlso se.ErrorCode = 10053) OrElse exc.Message = "Not connected to the server." Then ' Connection was aborted by the software
					Return True
				End If
			End If

			Return False
		End Function

		Public Sub ApplyLoginSettings(ByVal settings As SettingInfoBase) Implements IClientPlugin.ApplyLoginSettings
			Dim ftp As ComponentPro.Net.Ftp = CType(_client, ComponentPro.Net.Ftp)

			ftp.Passive = _loginSettings.Get(Of Boolean)(FtpLoginInfo.PasvMode)

			Dim proxyServer As String = _loginSettings.Get(Of String)(LoginInfo.ProxyServer)
			Dim proxyPort As Integer = _loginSettings.Get(Of Integer)(LoginInfo.ProxyPort)

			If (Not String.IsNullOrEmpty(proxyServer)) AndAlso proxyPort > 0 Then
				Dim proxy As New FtpProxy()
				ftp.Proxy = proxy

				proxy.Server = proxyServer
				proxy.Port = proxyPort
				proxy.UserName = _loginSettings.Get(Of String)(LoginInfo.ProxyUser)
				proxy.Password = _loginSettings.Get(Of String)(LoginInfo.ProxyPassword)
				proxy.Domain = _loginSettings.Get(Of String)(LoginInfo.ProxyDomain)
				proxy.ProxyType = _loginSettings.Get(Of FtpProxyType)(LoginInfo.ProxyType)
				proxy.AuthenticationMethod = _loginSettings.Get(Of FtpProxyHttpConnectAuthMethod)(LoginInfo.ProxyHttpAuthnMethod)
			End If
		End Sub

		Public Sub ApplySettings(ByVal settings As SettingInfoBase) Implements IClientPlugin.ApplySettings
			Dim client As ComponentPro.Net.Ftp = CType(_client, ComponentPro.Net.Ftp)

			client.Config.KeepAliveDuringIdleInterval = settings.Get(Of Integer)(SettingInfo.KeepAlive)
			If settings.Get(Of Boolean)(FtpSettingInfo.Compress) Then
				client.TransferMode = FtpTransferMode.ZlibCompressed
			Else
				client.TransferMode = FtpTransferMode.Stream
			End If
			client.Config.SendAbortCommand = settings.Get(Of OptionValue)(FtpSettingInfo.SendAborCommand)
			client.Config.SendTelnetInterruptSignal = settings.Get(Of Boolean)(FtpSettingInfo.SendAbortSignals)
			client.ChangeDirectoryBeforeFileOperation = settings.Get(Of Boolean)(FtpSettingInfo.ChangeDirBeforeTransfer)
			client.ChangeDirectoryBeforeListing = settings.Get(Of Boolean)(FtpSettingInfo.ChangeDirBeforeListing)
			client.Config.SmartPathResolving = settings.Get(Of Boolean)(FtpSettingInfo.SmartPath)
		End Sub

		''' <summary>
		''' Returns all issues of the given certificate.
		''' </summary>
		Private Shared Function GetCertProblem(ByVal status As CertificateVerificationStatus, ByVal code As Integer, ByRef showAddTrusted As Boolean) As String
			Select Case status
				Case CertificateVerificationStatus.TimeNotValid
					Return "Server's certificate has expired or is not valid yet."

				Case CertificateVerificationStatus.Revoked
					Return "Server's certificate has been revoked."

				Case CertificateVerificationStatus.UnknownCA
					Return "Server's certificate was issued by an unknown authority."

				Case CertificateVerificationStatus.RootNotTrusted
					showAddTrusted = True
					Return "Server's certificate was issued by an untrusted authority."

				Case CertificateVerificationStatus.IncompleteChain
					Return "Server's certificate does not chain up to a trusted root authority."

				Case CertificateVerificationStatus.Malformed
					Return "Server's certificate is malformed."

				Case CertificateVerificationStatus.CNNotMatch
					Return "Server hostname does not match the certificate."

				Case CertificateVerificationStatus.UnknownError
					Return String.Format("Error {0:x} encountered while validating server's certificate.", code)

				Case Else
					Return status.ToString()
			End Select
		End Function

		Private Sub client_CertificateReceived(ByVal sender As Object, ByVal e As ComponentPro.Security.CertificateReceivedEventArgs)
			Dim dlg As New CertValidator()

			Dim status As CertificateVerificationStatus = e.Status

			Dim values() As CertificateVerificationStatus = CType(System.Enum.GetValues(GetType(CertificateVerificationStatus)), CertificateVerificationStatus())

			Dim sbIssues As New StringBuilder()
			Dim showAddTrusted As Boolean = False
			For i As Integer = 0 To values.Length - 1
				' Matches the validation status?
				If (status And values(i)) = 0 Then
					Continue For
				End If

				' The issue is processed.
				status = status Xor values(i)

				sbIssues.AppendFormat("{0}" & vbCrLf, GetCertProblem(values(i), e.ErrorCode, showAddTrusted))
			Next i

			dlg.Certificate = e.ServerCertificates(0)
			dlg.Issues = sbIssues.ToString()
			dlg.ShowAddToTrustedList = showAddTrusted

			dlg.ShowDialog()

			e.AddToTrustedRoot = dlg.AddToTrustedList
			e.Accept = dlg.Accepted
		End Sub

		Private Sub client_CertificateRequired(ByVal sender As Object, ByVal e As ComponentPro.Security.CertificateRequiredEventArgs)
			Dim cert As String = _loginSettings.Get(Of String)(FtpLoginInfo.Certificate)

			' If the client cert file is specified.
			If Not String.IsNullOrEmpty(cert) Then
				' Load Certificate.
				Dim passdlg As New PasswordPrompt("Please provide password for certificate")
				' Ask for cert's passpharse
				If passdlg.ShowDialog() = System.Windows.Forms.DialogResult.OK Then
					Dim clientCert As New X509Certificate2(cert, passdlg.Password)
					e.Certificates = New X509Certificate2Collection(clientCert)
					Return
				End If

				' Password has not been provided.
			End If
			Dim dlg As New CertProvider()
			dlg.ShowDialog()
			' Get the selected certificate.
			e.Certificates = New X509Certificate2Collection(dlg.SelectedCertificate)
		End Sub

		''' <summary>
		''' Parses permissions from the given permissions string.
		''' </summary>
		''' <param name="data">The permissions string.</param>
		Private Shared Function ToPermissions(ByVal data As FtpFilePermissions) As String
			Return Convert.ToString(CUInt(data), 8) & "  " & FtpFileInfo.GetPermissionsString(data)
		End Function

		Public Sub Connect() Implements IClientPlugin.Connect
			_client.ConnectAsync(_loginSettings.Get(Of String)(LoginInfo.ServerName), _loginSettings.Get(Of Integer)(LoginInfo.ServerPort), _loginSettings.Get(Of FtpSecurityMode)(FtpLoginInfo.SecurityMode))
		End Sub

		Public Sub Authenticate() Implements IClientPlugin.Authenticate
			' Asynchronously login.
			_client.AuthenticateAsync(_loginSettings.Get(Of String)(LoginInfo.UserName), _loginSettings.Get(Of String)(LoginInfo.Password))
		End Sub

		Public Sub DoSetFilePermissions(ByVal files() As AbstractFileInfo, ByVal permissions As String, ByVal recursive As Boolean)
			Dim showProgress As Boolean = _controller.Settings.Get(Of Boolean)(SettingInfo.ShowProgressWhileDeleting)

			_controller.EnableProgress(True)
#If Framework4_5 OrElse (Not ASYNC) Then
			Try
#If Framework4_5 Then
				' Asynchronously set permissions of multiple files.
				If recursive Then
					Await client.SetMultipleFilesPermissionsAsync(_currentDirectory, files, permissions, showProgress,RecursionMode.RecurseIntoAllSubFolders, Nothing)
				Else
					Await client.SetMultipleFilesPermissionsAsync(_currentDirectory, files, permissions, showProgress,RecursionMode.None, Nothing)
				End If
#Else
				If recursive Then
					client.SetMultipleFilesPermissions(_currentDirectory, files, permissions, showProgress,RecursionMode.RecurseIntoAllSubFolders, Nothing)
				Else
					client.SetMultipleFilesPermissions(_currentDirectory, files, permissions, showProgress,RecursionMode.None, Nothing)
				End If
#End If
			Catch ex As Exception
				If Not HandleException(exc) Then
					Return
				End If
			End Try

			RefreshRemoteList()
			EnableProgress(False)
#Else
			If recursive Then
				_client.SetMultipleFilesPermissionsAsync(_controller.CurrentRemoteDirectory, files, permissions, showProgress,RecursionMode.RecurseIntoAllSubFolders, Nothing)
			Else
				_client.SetMultipleFilesPermissionsAsync(_controller.CurrentRemoteDirectory, files, permissions, showProgress,RecursionMode.None, Nothing)
			End If
#End If
		End Sub

		Public Sub DoLocalUploadUnique(ByVal localFile As String)
			_controller.EnableProgress(True)

#If Framework4_5 Then
			Try
				' Asynchronously upload unique the selected file.
				Await client.UploadUniqueFileAsync(localFile)
			Catch ex As Exception
				If Not HandleException(e.Error) Then
					Return
				End If
			End Try
#Else
			' Asynchronously upload unique the selected file.
			_client.UploadUniqueFileAsync(localFile)
#End If
		End Sub

		''' <summary>
		''' Handles the client's UploadUniqueCompleted event.
		''' </summary>
		''' <param name="sender">The Ftp object.</param>
		''' <param name="e">The event arguments.</param>
		Private Sub client_UploadUniqueFileCompleted(ByVal sender As Object, ByVal e As ExtendedAsyncCompletedEventArgs(Of String))
			If e.Error IsNot Nothing Then
				If Not _controller.HandleException(e.Error) Then
					Return
				End If
			End If

			_controller.EnableProgress(False)
			_controller.RefreshRemoteList()
		End Sub
	End Class
End Namespace

File: FTP Client for WinForms using Ultimate FTP library/[C#]-FTP Client for WinForms using Ultimate FTP library/C#/Ftp/FtpClientPlugin.cs
using System;
using System.Collections.Generic;
using System.Security.Cryptography.X509Certificates;
using System.Text;
using ClientSample.Ftp.Security;
using ComponentPro;
using ComponentPro.IO;
using ComponentPro.Net;
using ComponentPro.Security.Certificates;

namespace ClientSample.Ftp
{
    class FtpClientPlugin : IClientPlugin
    {
        private ClientController _controller;
        private ComponentPro.Net.Ftp _client;
        private IClientView _view;
        private SettingInfoBase _loginSettings;

        public FtpClientPlugin(IClientView view, ClientController clientController, SettingInfoBase loginSettings)
        {
            _view = view;
            _controller = clientController;
            _loginSettings = loginSettings;
        }

        public IRemoteFileSystem Create()
        {
            ComponentPro.Net.Ftp client = new ComponentPro.Net.Ftp();

#if !Framework4_5
            client.SetMultipleFilesPermissionsCompleted += client_SetMultipleFilesPermissionsCompleted;
            client.UploadUniqueFileCompleted += client_UploadUniqueFileCompleted;
            client.CertificateRequired += client_CertificateRequired;
            client.CertificateReceived += client_CertificateReceived;
#endif

            _client = client;

            return _client;
        }
        
        /// <summary>
        /// Handles the client's SetMultipleFilesPermissionsCompleted event.
        /// </summary>
        /// <param name="sender">The Ftp object.</param>
        /// <param name="e">The event arguments.</param>
        void client_SetMultipleFilesPermissionsCompleted(object sender, ExtendedAsyncCompletedEventArgs<FileSystemTransferStatistics> e)
        {
            if (e.Error != null)
            {
                if (!_controller.HandleException(e.Error))
                    return;
            }

            _controller.RefreshRemoteList();
            _controller.EnableProgress(false);
        }

        public string GetPermissionsString(AbstractFileInfo f)
        {
            if (((FtpFileInfo)f).System == FtpFileOs.Unix)
                return ToPermissions(((FtpFileInfo)f).Permissions);
            else
                return "-";
        }

        void IClientPlugin.AuthenticatePost()
        {
            if (_loginSettings.Get<bool>(FtpLoginInfo.ClearCommandChannel) && _loginSettings.Get<FtpSecurityMode>(FtpLoginInfo.SecurityMode) != FtpSecurityMode.None)
            {
                ((ComponentPro.Net.Ftp)_client).ClearCommandChannel();
            }
        }

        public bool CanBeReconnected(Exception exc)
        {
            FtpException fe = exc as FtpException;
            System.Net.Sockets.SocketException se;

            if (fe != null)
            {
                if (fe.Status == FtpExceptionStatus.ConnectionClosed)
                {
                    return true;
                }
            }
            else
            {
                se = exc as System.Net.Sockets.SocketException;
                if (
                    (se != null && se.ErrorCode == 10053) || // Connection was aborted by the software
                    exc.Message == "Not connected to the server."
                    )
                {
                    return true;
                }
            }

            return false;
        }

        public void ApplyLoginSettings(SettingInfoBase settings)
        {
            ComponentPro.Net.Ftp ftp = (ComponentPro.Net.Ftp)_client;

            ftp.Passive = _loginSettings.Get<bool>(FtpLoginInfo.PasvMode);

            string proxyServer = _loginSettings.Get<string>(LoginInfo.ProxyServer);
            int proxyPort = _loginSettings.Get<int>(LoginInfo.ProxyPort);

            if (!string.IsNullOrEmpty(proxyServer) && proxyPort > 0)
            {
                FtpProxy proxy = new FtpProxy();
                ftp.Proxy = proxy;

                proxy.Server = proxyServer;
                proxy.Port = proxyPort;
                proxy.UserName = _loginSettings.Get<string>(LoginInfo.ProxyUser);
                proxy.Password = _loginSettings.Get<string>(LoginInfo.ProxyPassword);
                proxy.Domain = _loginSettings.Get<string>(LoginInfo.ProxyDomain);
                proxy.ProxyType = _loginSettings.Get<FtpProxyType>(LoginInfo.ProxyType);
                proxy.AuthenticationMethod = _loginSettings.Get<FtpProxyHttpConnectAuthMethod>(LoginInfo.ProxyHttpAuthnMethod);
            }
        }

        public void ApplySettings(SettingInfoBase settings)
        {
            ComponentPro.Net.Ftp client = (ComponentPro.Net.Ftp) _client;

            client.Config.KeepAliveDuringIdleInterval = settings.Get<int>(SettingInfo.KeepAlive);
            client.TransferMode = settings.Get<bool>(FtpSettingInfo.Compress) ? FtpTransferMode.ZlibCompressed : FtpTransferMode.Stream;
            client.Config.SendAbortCommand = settings.Get<OptionValue>(FtpSettingInfo.SendAborCommand);
            client.Config.SendTelnetInterruptSignal = settings.Get<bool>(FtpSettingInfo.SendAbortSignals);
            client.ChangeDirectoryBeforeFileOperation = settings.Get<bool>(FtpSettingInfo.ChangeDirBeforeTransfer);
            client.ChangeDirectoryBeforeListing = settings.Get<bool>(FtpSettingInfo.ChangeDirBeforeListing);
            client.Config.SmartPathResolving = settings.Get<bool>(FtpSettingInfo.SmartPath);
        }

        /// <summary>
        /// Returns all issues of the given certificate.
        /// </summary>
        private static string GetCertProblem(CertificateVerificationStatus status, int code, ref bool showAddTrusted)
        {
            switch (status)
            {
                case CertificateVerificationStatus.TimeNotValid:
                    return "Server's certificate has expired or is not valid yet.";

                case CertificateVerificationStatus.Revoked:
                    return "Server's certificate has been revoked.";

                case CertificateVerificationStatus.UnknownCA:
                    return "Server's certificate was issued by an unknown authority.";

                case CertificateVerificationStatus.RootNotTrusted:
                    showAddTrusted = true;
                    return "Server's certificate was issued by an untrusted authority.";

                case CertificateVerificationStatus.IncompleteChain:
                    return "Server's certificate does not chain up to a trusted root authority.";

                case CertificateVerificationStatus.Malformed:
                    return "Server's certificate is malformed.";

                case CertificateVerificationStatus.CNNotMatch:
                    return "Server hostname does not match the certificate.";

                case CertificateVerificationStatus.UnknownError:
                    return string.Format("Error {0:x} encountered while validating server's certificate.", code);

                default:
                    return status.ToString();
            }
        }

        void client_CertificateReceived(object sender, ComponentPro.Security.CertificateReceivedEventArgs e)
        {
            CertValidator dlg = new CertValidator();

            CertificateVerificationStatus status = e.Status;

            CertificateVerificationStatus[] values = (CertificateVerificationStatus[])Enum.GetValues(typeof(CertificateVerificationStatus));

            StringBuilder sbIssues = new StringBuilder();
            bool showAddTrusted = false;
            for (int i = 0; i < values.Length; i++)
            {
                // Matches the validation status?
                if ((status & values[i]) == 0)
                    continue;

                // The issue is processed.
                status ^= values[i];

                sbIssues.AppendFormat("{0}\r\n", GetCertProblem(values[i], e.ErrorCode, ref showAddTrusted));
            }

            dlg.Certificate = e.ServerCertificates[0];
            dlg.Issues = sbIssues.ToString();
            dlg.ShowAddToTrustedList = showAddTrusted;

            dlg.ShowDialog();

            e.AddToTrustedRoot = dlg.AddToTrustedList;
            e.Accept = dlg.Accepted;
        }

        private void client_CertificateRequired(object sender, ComponentPro.Security.CertificateRequiredEventArgs e)
        {
            string cert = _loginSettings.Get<string>(FtpLoginInfo.Certificate);
            
            // If the client cert file is specified.
            if (!string.IsNullOrEmpty(cert))
            {
                // Load Certificate.
                PasswordPrompt passdlg = new PasswordPrompt("Please provide password for certificate");
                // Ask for cert's passpharse
                if (passdlg.ShowDialog() == System.Windows.Forms.DialogResult.OK)
                {
                    X509Certificate2 clientCert = new X509Certificate2(cert, passdlg.Password);
                    e.Certificates = new X509Certificate2Collection(clientCert);
                    return;
                }

                // Password has not been provided.
            }
            CertProvider dlg = new CertProvider();
            dlg.ShowDialog();
            // Get the selected certificate.
            e.Certificates = new X509Certificate2Collection(dlg.SelectedCertificate);
        }

        /// <summary>
        /// Parses permissions from the given permissions string.
        /// </summary>
        /// <param name="data">The permissions string.</param>
        private static string ToPermissions(FtpFilePermissions data)
        {
            return Convert.ToString((uint)data, 8) + "  " + FtpFileInfo.GetPermissionsString(data);
        }

        public void Connect()
        {
            _client.ConnectAsync(_loginSettings.Get<string>(LoginInfo.ServerName), _loginSettings.Get<int>(LoginInfo.ServerPort),
                                 _loginSettings.Get<FtpSecurityMode>(FtpLoginInfo.SecurityMode));
        }

        public void Authenticate()
        {
            // Asynchronously login.
            _client.AuthenticateAsync(_loginSettings.Get<string>(LoginInfo.UserName), _loginSettings.Get<string>(LoginInfo.Password));
        }

        public void DoSetFilePermissions(AbstractFileInfo[] files, string permissions, bool recursive)
        {
            bool showProgress = _controller.Settings.Get<bool>(SettingInfo.ShowProgressWhileDeleting);

            _controller.EnableProgress(true);
#if Framework4_5 || !ASYNC
            try
            {
#if Framework4_5
                // Asynchronously set permissions of multiple files.
                await client.SetMultipleFilesPermissionsAsync(_currentDirectory, files, permissions, showProgress, recursive ? RecursionMode.RecurseIntoAllSubFolders : RecursionMode.None, null);
#else
                client.SetMultipleFilesPermissions(_currentDirectory, files, permissions, showProgress, recursive ? RecursionMode.RecurseIntoAllSubFolders : RecursionMode.None, null);
#endif
            }
            catch (Exception ex)
            {
                if (!HandleException(exc))
                    return;
            }

            RefreshRemoteList();
            EnableProgress(false);
#else
            _client.SetMultipleFilesPermissionsAsync(_controller.CurrentRemoteDirectory, files, permissions, showProgress, recursive ? RecursionMode.RecurseIntoAllSubFolders : RecursionMode.None, null);
#endif
        }

        public void DoLocalUploadUnique(string localFile)
        {
            _controller.EnableProgress(true);

#if Framework4_5
            try
            {
                // Asynchronously upload unique the selected file.
                await client.UploadUniqueFileAsync(localFile);
            }
            catch (Exception ex)
            {
                if (!HandleException(e.Error))
                    return;
            }
#else
            // Asynchronously upload unique the selected file.
            _client.UploadUniqueFileAsync(localFile);
#endif
        }

        /// <summary>
        /// Handles the client's UploadUniqueCompleted event.
        /// </summary>
        /// <param name="sender">The Ftp object.</param>
        /// <param name="e">The event arguments.</param>
        private void client_UploadUniqueFileCompleted(object sender, ExtendedAsyncCompletedEventArgs<string> e)
        {
            if (e.Error != null)
            {
                if (!_controller.HandleException(e.Error))
                    return;
            }

            _controller.EnableProgress(false);
            _controller.RefreshRemoteList();
        }
    }
}
